bitcoin/src/taproot.rs

```diff
@@ -521,26 +576,238 @@ impl Default for TaprootBuilder {
     fn default() -> Self { Self::new() }
 }
 
-/// Represents the node information in taproot tree.
+/// Error happening when [`TapTree`] is constructed from a [`TaprootBuilder`]
+/// having hidden branches or not being finalized.
+#[derive(Clone, Ord, PartialOrd, Eq, PartialEq, Hash, Debug)]
+#[non_exhaustive]
+pub enum IncompleteBuilder {
+    /// Indicates an attempt to construct a tap tree from a builder containing incomplete branches.
+    NotFinalized(TaprootBuilder),
+    /// Indicates an attempt to construct a tap tree from a builder containing hidden parts.
+    HiddenParts(TaprootBuilder),
+}
+
+impl IncompleteBuilder {
+    /// Converts error into the original incomplete [`TaprootBuilder`] instance.
+    pub fn into_builder(self) -> TaprootBuilder {
+        match self {
+            IncompleteBuilder::NotFinalized(builder) | IncompleteBuilder::HiddenParts(builder) =>
+                builder,
+        }
+    }
+}
+
+impl core::fmt::Display for IncompleteBuilder {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.write_str(match self {
+            IncompleteBuilder::NotFinalized(_) =>
+                "an attempt to construct a tap tree from a builder containing incomplete branches.",
+            IncompleteBuilder::HiddenParts(_) =>
+                "an attempt to construct a tap tree from a builder containing hidden parts.",
+        })
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+impl std::error::Error for IncompleteBuilder {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use self::IncompleteBuilder::*;
```

Comment:
OT: These `use self::<ErrorType>::*` statements are sprinkled around the codebase because I was coding sometime with only half my brain connected, the `self::` is totally useless.